### PR TITLE
vxlan_vtep and vxlan_vtep_vni refactoring

### PR DESCRIPTION
**Pull request by mikewiebe**
**Summary of changes**
- vxlan_vtep
  - Renamed mac_distribution property to be host_reachability and simplifed.
  - Added shutdown yaml entry.
  - Removed dead code
  - Added minitests for new code
- vxlan_vtep_vni
  - Added associate-vrf parameter
  - Added minitests for now code

**MiniTest Results**

```
rtp-ads-432:/nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils> ruby tests/test_vxlan_vtep.rb -v -- 10.122.197.222 admin admin
Run options: -v -- --seed 59261

# Running:


Node under test:
  - name  - dt-n9k5-1
  - type  - N9K-C9396PX
  - image - bootflash:///nxos.7.0.3.I2.1.74.bin

TestVxlanVtep#test_shutdown = 16.55 s = .
TestVxlanVtep#test_description = 9.72 s = .
TestVxlanVtep#test_source_interface = 10.12 s = .
TestVxlanVtep#test_create_multiple_negative = 8.46 s = .
TestVxlanVtep#test_create_destroy_one = 9.08 s = .
TestVxlanVtep#test_host_reachability = 9.49 s = .
TestVxlanVtep#test_create_destroy_multiple = 0.83 s = S

Finished in 64.250164s, 0.1089 runs/s, 0.3424 assertions/s.

  1) Skipped:
TestVxlanVtep#test_create_destroy_multiple [tests/test_vxlan_vtep.rb:53]:
Only supported on n7k

7 runs, 22 assertions, 0 failures, 0 errors, 1 skips
Coverage report generated for MiniTest to /nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils/coverage. 446 / 526 LOC (84.79%) covered.
rtp-ads-432:/nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils> 
```

```
rtp-ads-432:/nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils> ruby tests/test_vxlan_vtep_vni.rb -v -- 10.122.197.222 admin admin
Run options: -v -- --seed 46655

# Running:


Node under test:
  - name  - dt-n9k5-1
  - type  - N9K-C9396PX
  - image - bootflash:///nxos.7.0.3.I2.1.74.bin

TestVxlanVtepVni#test_suppress_arp = 9.99 s = E
TestVxlanVtepVni#test_ingress_replication = 11.55 s = .
TestVxlanVtepVni#test_multicast_group = 11.00 s = .
TestVxlanVtepVni#test_create_with_existing = 10.24 s = .
TestVxlanVtepVni#test_peer_list = 10.81 s = .
TestVxlanVtepVni#test_vnis = 12.10 s = .

Finished in 65.698252s, 0.0913 runs/s, 0.6697 assertions/s.

  1) Error:
TestVxlanVtepVni#test_suppress_arp:
Cisco::CliError: CliError: ' suppress-arp' rejected with message:
'ERROR: Please configure TCAM region for Ingress ARP-Ether ACL before configuring ARP supression.'
    /nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils/lib/cisco_node_utils/node.rb:266:in `rescue in config'
    /nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils/lib/cisco_node_utils/node.rb:263:in `config'
    /nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils/lib/cisco_node_utils/node.rb:161:in `config_set'
    /nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils/lib/cisco_node_utils/node_util.rb:58:in `config_set'
    /nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils/lib/cisco_node_utils/vxlan_vtep_vni.rb:204:in `suppress_arp='
    tests/test_vxlan_vtep_vni.rb:170:in `test_suppress_arp'

6 runs, 44 assertions, 0 failures, 1 errors, 0 skips
Coverage report generated for MiniTest to /nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils/coverage. 540 / 681 LOC (79.3%) covered.
rtp-ads-432:/nobackup/mwiebe/REPOS_NODE_UTILS/vxlan/cisco-network-node-utils> 
```

**NOTE:** Suppress arp error is expected... box requires a reload to set the prop.

### CODE REVIEW

File: lib/cisco_node_utils/vxlan_vtep.rb [hunk 1]
```diff
@@ -90,121 +90,48 @@ def description=(desc)
...
+      hr == 'bgp' ? 'evpn' : hr
+    end
+
+    def host_reachability=(val)
+      set_args = { name: @name, proto: 'bgp' }
+      if val.to_s == 'flood' && host_reachability == 'evpn'
+        set_args[:state] = 'no'
+      elsif val.to_s == 'evpn'
+        set_args[:state] = ''
       end
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep.rb]: If neither if condition is true then `:state` will not be set when you call config_set. Is that okay?

File: lib/cisco_node_utils/vxlan_vtep.rb [hunk 2]
```diff
@@ -90,121 +90,48 @@ def description=(desc)
...
     def source_interface
-      src_intf = config_get('vxlan_vtep', 'source_intf', name: @name)
-      return default_source_interface if src_intf.nil?
-      src_intf
+      config_get('vxlan_vtep', 'source_intf', name: @name)
+    end
+
+    def source_interface_set(val)
+      set_args = { name: @name, lpbk_intf: val }
+      val.empty? ? set_args[:state] = 'no' : set_args[:state] = ''
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep.rb]: nit:
`set_args[:state] = val.empty? ? 'no' : ''`

File: lib/cisco_node_utils/vxlan_vtep.rb [hunk 3]
```diff
@@ -90,121 +90,48 @@ def description=(desc)
...
-        config_set('vxlan_vtep', 'source_intf',
-                   name: @name, state: '', lpbk_intf: val)
-      end
-    rescue Cisco::CliError => e
-      raise "[#{@name}] '#{e.command}' : #{e.clierror}"
+      # The source interface can only be changed if the nve
+      # interface is in a shutdown state.
+      current_state = shutdown
+      self.shutdown = true unless shutdown
+      source_interface_set(val)
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep.rb:133]: This is okay but you only call `source_interface_set` from one place. You could just move the method contents into this setter itself.

File: lib/cisco_node_utils/vxlan_vtep_vni.rb [hunk 4]
```diff
@@ -24,13 +24,19 @@
 module Cisco
   # VxlanVtepVni - node utility for vxlan vtep vni members.
   class VxlanVtepVni < NodeUtil
-    attr_reader :name, :vni
+    attr_reader :name, :vni, :vrf
 
-    def initialize(name, vni, instantiate=true)
+    def initialize(name, vni, vrf=false, instantiate=true)
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: nit: I would use `assoc_vrf` here, only because vrf means something completely different in all of our other providers. Your call.

File: lib/cisco_node_utils/vxlan_vtep_vni.rb [hunk 5]
```diff
@@ -24,13 +24,19 @@
...
       @name = name
       @vni = vni
+      @vrf = vrf
 
       create if instantiate
+      # NOTE: This can be removed after the puppet type/provider are coded
+      # but a few design notes so we don't forget!.
+      # 1) The title pattern should be cisco_vxlan_vtep_vni { 'name', 'vni' :
+      # 2) vrf will be a property but will also be a namevar.
+      # 3) The type must check that vrf is always set to either true or false.
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: Good idea. But now you need a breadcrumb in the puppet code to remember to remove this breadcrumb. :-)

File: lib/cisco_node_utils/vxlan_vtep_vni.rb [hunk 6]
```diff
@@ -40,19 +46,21 @@ def self.vnis
         get_args = { name: name }
         vni_list = config_get('vxlan_vtep_vni', 'all_vnis', get_args)
         next if vni_list.nil?
-        vni_list.each do |vni|
-          hash[name][vni] = VxlanVtepVni.new(name, vni, false)
+        vni_list.each do |vni, vrf|
+          vrf.nil? ? vrf = false : vrf = true
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: nit: `vrf = vrf.nil? ? false : true`

File: lib/cisco_node_utils/vxlan_vtep_vni.rb [hunk 7]
```diff
@@ -40,19 +46,21 @@ def self.vnis
...
     end
 
     def ==(other)
-      (name == other.name) && (vni == other.vni)
+      (name == other.name) && (vni == other.vni) && (vrf == other.vrf)
     end
 
     def set_args_keys_default
       keys = { name: @name, vni: @vni }
+      @vrf ? keys[:vrf] = 'associate-vrf' : keys[:vrf] = ''
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: ditto nit: tertiary format

File: lib/cisco_node_utils/vxlan_vtep_vni.rb [hunk 8]
```diff
@@ -63,8 +71,34 @@ def set_args_keys(hash={})
...
+      # - member nve 5000 associate-vrf
+      # They are mutually exclusive and one must be removed before the other
+      # can be configured.
       set_args_keys(state: '')
+      if create_with_associate_vrf?
+        destroy_existing('vni_without_vrf')
+      else
+        destroy_existing('vni_with_vrf')
+      end
       config_set('vxlan_vtep', 'vni', @set_args)
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: nit: you can remove the `set_args_keys(...` call above and make this line:

`    config_set('vxlan_vtep', 'vni', set_args_keys(state: ''))`

Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb:99]: nit: you can remove the set_args_keys(... call above and make this line:

config_set('vxlan_vtep', 'vni', set_args_keys(state: ''))

File: lib/cisco_node_utils/vxlan_vtep_vni.rb [hunk 9]
```diff
@@ -63,8 +71,34 @@ def set_args_keys(hash={})
...
+      getargs = { name: @name, vni: @vni, state: '' }
+      # rubocop:disable Style/GuardClause
+      if config_get('vxlan_vtep', key, getargs)
+        key.eql?('vni_with_vrf') ? vrf = 'associate-vrf' : vrf = ''
+        getargs[:vrf] = vrf
+        getargs[:state] = 'no'
+        config_set('vxlan_vtep', 'vni', getargs)
+      end
+      # rubocop:enable Style/GuardClause
+    end
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb:85]: Why bypass the guard clause?  What's wrong with this:

```
    def destroy_existing(key)
      getargs = { name: @name, vni: @vni, state: '' }
      return unless config_get('vxlan_vtep', key, getargs)
      getargs[:vrf] = key[/vni_with_vrf/] ? 'associate-vrf' : ''
      getargs[:state] = 'no'
      config_set('vxlan_vtep', 'vni', getargs)
    end
```

File: tests/test_vxlan_vtep.rb [hunk 10]
```diff
@@ -110,24 +110,21 @@ def test_description
...
+  def test_host_reachability
     vtep = VxlanVtep.new('nve1')
 
-    val = :flood
-    vtep.mac_distribution = val
-    assert_equal(vtep.mac_distribution, val.to_s,
-                 "mac_distribution is not #{val}")
+    vtep.host_reachability = 'flood'
+    assert_equal(vtep.host_reachability, 'flood',
+                 "host_reachability is not 'flood'")
```
Comment (chrisvanheuveln) [inline tests/test_vxlan_vtep.rb]: fyi: I tend to not include an error message for `assert_equal` calls since on failure they'll display `test_host_reachability` `Expected: 'flood'` `Found: 'foobar'` so the message usually doesn't add much value. But it's fine as is.

File: lib/cisco_node_utils/vxlan_vtep.rb (hunk 1 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep.rb:110]: Yes.. that is the intent.  If we want to set host_reachability to 'flood' and the current value setting is flood it becomes a no-op.

File: lib/cisco_node_utils/vxlan_vtep.rb (hunk 3 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep.rb:133]: Muddies the waters too much for my taste.  IMO it's easier to understand the shut/noshut logic without the extra unrelated code.

File: lib/cisco_node_utils/vxlan_vtep.rb (hunk 2 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep.rb]: changed

File: lib/cisco_node_utils/vxlan_vtep_vni.rb (hunk 4 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: Yeah.. glad you made the comment.  I went aback and forth with several names before I settled on `vrf` but based on your suggestion I agree that `assoc_vrf` adds more clarity.

File: lib/cisco_node_utils/vxlan_vtep_vni.rb (hunk 6 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: ack

File: lib/cisco_node_utils/vxlan_vtep_vni.rb (hunk 7 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: ditto ack.

File: lib/cisco_node_utils/vxlan_vtep_vni.rb (hunk 9 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb:85]: Because it was late and I was tired and it annoyed me :)  I'll update with your suggestion.

File: lib/cisco_node_utils/vxlan_vtep_vni.rb (hunk 8 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb:99]: I actually can't.... :) I need to call `set_args_keys` before I call `create_with_associate_vrf?` so that @set_args has stuff in it.  Alternatively I could just get rid of `create_with_associate_vrf` but that was slightly more readable then `!@set_args[:vrf].eql?('')`

File: tests/test_vxlan_vtep.rb (hunk 10 above)
Comment (mikewiebe) [inline tests/test_vxlan_vtep.rb]: True.. I have done that in several cases as well.. I'll remove them

File: lib/cisco_node_utils/vxlan_vtep_vni.rb (hunk 5 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb]: We don't have any puppet code for this yet and I figured whoever works on puppet will likely work with the nodeutils code as well and very likely see this.  I was supposed to work on the puppet side of this next but if you get to it first... consider yourself warned :)

**Comment by chrisvanheuveln** [comment]
I forgot to comment earlier on the suppress-arp test failure. You should add a rescue to the test case so that it skips if it detects the TCAM error message, and the skip message can indicate why it was skipped.

File: lib/cisco_node_utils/vxlan_vtep.rb (hunk 3 above)
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep.rb:133]: yup, that's fine then.

File: lib/cisco_node_utils/vxlan_vtep_vni.rb (hunk 8 above)
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/vxlan_vtep_vni.rb:99]: It's fine. Leave it.
